akka-persistence-testkit/src/main/scala/akka/persistence/testkit/internal/TestKitStorage.scala

```diff
@@ -47,15 +48,27 @@ sealed trait InMemStorage[K, R] extends InternalReprSupport[R] {
           Some(value.drop(fromInclusive).take(maxNum))
         else None)
 
+  def removeFirstInExpectNextQueue(key: K): Unit = synchronized(expectNextQueue.get(key).foreach(_.poll()))
+
+  def firstInExpectNextQueue(key: K): Option[R] = expectNextQueue.get(key).flatMap { item =>
+    Try(item.element()).toOption.map(toRepr)
+  }
+
   def findOneByIndex(key: K, index: Int): Option[R] =
-    Option(eventsMap.get(key))
+    eventsMap
+      .get(key)
       .flatMap {
         case (_, value) => if (value.size > index) Some(value(index)) else None
       }
       .map(toRepr)
 
-  def add(key: K, p: R): Unit =
+  def add(key: K, p: R): Unit = synchronized {
```

Comment:
Since this is a trait we are not in full control of `this` and therefore I'd suggest that we synchronize on a separate `lock` rather than `this`.

```
// synchronized access to the state here
private val lock = new Object

def add(key: K, p: R): Unit = lock.synchronized {
```

Reply:
👍